### PR TITLE
Implemented chmod

### PR DESCRIPTION
**Pull request by Dennisbonke**
This PR implements a `FCHMODAT` request in posix, allowing the libc functions `chmod()`, `fchmod()` and `fchmodat()` to function.

### CODE REVIEW

File: drivers/libblockfs/src/libblockfs.cpp [hunk 1]
```diff
@@ -274,7 +274,7 @@ getStats(std::shared_ptr<void> object) {
 	protocols::fs::FileStats stats;
 	stats.linkCount = self->numLinks;
 	stats.fileSize = self->fileSize();
-	stats.mode = self->mode;
+	stats.mode = self->diskInode()->mode;
```
Comment (avdgrinten) [inline drivers/libblockfs/src/libblockfs.cpp:277]: We should remove `self->mode` entirely, in favor of `diskInode()->mode`.

Comment (avdgrinten) [reply drivers/libblockfs/src/libblockfs.cpp:277]: Also, when returning the mode here, we should mask out the file type.

File: posix/subsystem/src/fs.cpp [hunk 2]
```diff
@@ -83,6 +83,10 @@ DeviceId FsNode::readDevice() {
 	throw std::runtime_error("readDevice() is not implemented for this FsNode");
 }
 
+async::result<int> FsNode::chmod(int mode) {
+	throw std::runtime_error("chmod() is not implemented for this FsNode");
+}
```
Comment (avdgrinten) [inline posix/subsystem/src/fs.cpp:88]: Hm, we should return a `async::result<Error>` and handle this error gracefully. Probably by returning the equivalent of EPERM.

File: posix/subsystem/src/main.cpp [hunk 3]
```diff
@@ -1476,6 +1476,68 @@ async::result<void> serveRequests(std::shared_ptr<Process> self,
...
+					continue;
+				}
+
+				relative_to = {file->associatedMount(), file->associatedLink()};
+			}
+
+			if(req.flags()) {
+				if(req.flags() & AT_SYMLINK_NOFOLLOW) {
+					co_await sendErrorResponse(managarm::posix::Errors::NOT_SUPPORTED);
+					continue;
```
Comment (avdgrinten) [inline posix/subsystem/src/main.cpp:1510]: What does this translate to on the client side?

Comment (Dennisbonke) [reply posix/subsystem/src/main.cpp:1510]: ENOTSUP, which is the error returned in this case according to the man page

File: posix/subsystem/src/fs.cpp (hunk 2 above)
Comment (Dennisbonke) [reply posix/subsystem/src/fs.cpp:88]: Personally, I'd do ENOTSUP, as it is unsupported

File: drivers/libblockfs/src/ext2fs.cpp [hunk 4]
```diff
@@ -601,7 +614,7 @@ async::detached FileSystem::initiateInode(std::shared_ptr<Inode> inode) {
 
 	// filter out the file type from the mode
 	// TODO: ext2fs stores a 32-bit mode
-	inode->mode = disk_inode->mode & 0x0FFF;
+	inode->diskInode()->mode = disk_inode->mode & 0x0FFF;
```
Comment (avdgrinten) [inline drivers/libblockfs/src/ext2fs.cpp:617]: This line is not needed. (Before the change, it copied the mode to a member variable. Now it copies the mode to itself and discards the file type -- that's problematic.)

File: posix/subsystem/src/extern_fs.cpp [hunk 5]
```diff
@@ -74,6 +74,32 @@ struct Node : FsNode {
...
+
+		managarm::fs::CntRequest req;
+		req.set_req_type(managarm::fs::CntReqType::NODE_CHMOD);
+		req.set_mode(mode);
+
+		auto ser = req.SerializeAsString();
+		auto &&transmit = helix::submitAsync(getLane(), helix::Dispatcher::global(),
+				helix::action(&offer, kHelItemAncillary),
+				helix::action(&send_req, ser.data(), ser.size(), kHelItemChain),
+				helix::action(&recv_resp));
```
Comment (avdgrinten) [inline posix/subsystem/src/extern_fs.cpp:90]: Use helix_ng here.

File: posix/subsystem/src/fs.cpp [hunk 6]
```diff
@@ -83,6 +83,10 @@ DeviceId FsNode::readDevice() {
 	throw std::runtime_error("readDevice() is not implemented for this FsNode");
 }
 
+async::result<Error> FsNode::chmod(int mode) {
+	co_return Error::accessDenied;
```
Comment (avdgrinten) [inline posix/subsystem/src/fs.cpp:87]: This change is good. But we should maybe still log about this.

File: posix/subsystem/src/main.cpp (hunk 3 above)
Comment (avdgrinten) [reply posix/subsystem/src/main.cpp:1508]: OK!

File: protocols/fs/include/protocols/fs/server.hpp [hunk 7]
```diff
@@ -199,6 +199,8 @@ struct NodeOperations {
 	async::result<std::string> (*readSymlink)(std::shared_ptr<void> object);
 
 	async::result<MkdirResult> (*mkdir)(std::shared_ptr<void> object, std::string name);
+
+	async::result<int> (*chmod)(std::shared_ptr<void> object, int mode);
```
Comment (avdgrinten) [inline protocols/fs/include/protocols/fs/server.hpp:203]: Does protocols/fs have an `Error` enum? If yes, return that instead of int.

**Comment by qookei** [comment]
LGTM

**Pull request merged**